Funcky/DataTypes/EitherOrBoth.cs

```diff
@@ -0,0 +1,109 @@
+        }
+
+        private enum Side : byte
+        {
+            Uninitialized,
+            Both,
+            Left,
+            Right,
+        }
+
+        [Pure]
+        public static bool operator ==(EitherOrBoth<TLeft, TRight> lhs, EitherOrBoth<TLeft, TRight> rhs) => lhs.Equals(rhs);
+
+        [Pure]
+        public static bool operator !=(EitherOrBoth<TLeft, TRight> lhs, EitherOrBoth<TLeft, TRight> rhs) => !lhs.Equals(rhs);
+
+        [Pure]
+        public static EitherOrBoth<TLeft, TRight> Both(TLeft left, TRight right) => new (left, right);
+
+        [Pure]
+        public static EitherOrBoth<TLeft, TRight> Left(TLeft left) => new (left);
+
+        [Pure]
+        public static EitherOrBoth<TLeft, TRight> Right(TRight right) => new (right);
+
+        [Pure]
+        public TMatchResult Match<TMatchResult>(Func<TLeft, TMatchResult> left, Func<TRight, TMatchResult> right, Func<TLeft, TRight, TMatchResult> both)
+            => _side switch
+            {
+                Side.Left => left(_left),
+                Side.Right => right(_right),
+                Side.Both => both(_left, _right),
+                Side.Uninitialized => throw new NotSupportedException(
+                    "EitherOrBoth constructed via default instead of a factory function (EitherOrBoth.Left, EitherOrBoth.Right or EitherOrBoth.Both)"),
+                _ => throw new ArgumentOutOfRangeException(nameof(_side), $"Internal error: Enum variant {_side} is not handled"),
+            };
+
+        public void Match(Action<TLeft> left, Action<TRight> right, Action<TLeft, TRight> both)
+        {
+            switch (_side)
+            {
+                case Side.Left:
+                    left(_left);
+                    break;
+                case Side.Both:
+                    right(_right);
+                    break;
+                case Side.Right:
+                    both(_left, _right);
+                    break;
+                case Side.Uninitialized:
+                    throw new NotSupportedException("EitherOrBoth constructed via default instead of a factory function (EitherOrBoth.Left, EitherOrBoth.Right or EitherOrBoth.Both)");
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(_side), $"Internal error: Enum variant {_side} is not handled");
+            }
+        }
+
+        [Pure]
+        public override bool Equals(object? obj)
+            => obj is EitherOrBoth<TLeft, TRight> other && Equals(other);
+
+        [Pure]
+        public bool Equals(EitherOrBoth<TLeft, TRight> other)
+            => Equals(_side, other._side)
+               && Equals(_right, other._right)
+               && Equals(_left, other._left);
+
+        [Pure]
+        public override int GetHashCode()
+            => Match(
+                left => left?.GetHashCode(),
+                right => right?.GetHashCode(),
+                (left, right) => left?.GetHashCode() ^ right?.GetHashCode()) ?? 0;
```

Comment:
There's `HashCode.Combine`:
```suggestion
                (left, right) => HashCode.Combine(left, right)) ?? 0;
```

Reply:
HashCode.Combine is not supported in .net standard 2.9

Reply:
There’s a NuGet package for that: https://www.nuget.org/packages/Microsoft.Bcl.HashCode/

Reply:
Adressed in 879228c1